### PR TITLE
Don't rely on inlining observations if VM doesn't allow that

### LINKED ISSUE #95815: [R2R] Drop inlining observations when inlining is not performed
Reported by internal Microsoft team. Internal link: [Watson crash from hitting breakpoint?](https://watsonportal.microsoft.com/Failure?FailureSearchText=4da65c29-ed31-33e0-acb6-11c6b62ec228)

## Observed behavior

R2R images created with `--opt-cross-module:*` option have spurious `int3` instructions.

## Analysis of the problem

Inlining in RyuJIT is a two-step process:

-	Creating observations about the inline
-	Actual inlining

We record the dependency on the exact version of the inlined method body only when we perform the actual inlining. I do not see any code for recording dependency on observations.

The problem is that the optimizations can use results from observations even when the inlining is not performed. The observation causing problem here is [CALLEE_DOES_NOT_RETURN](https://github.com/search?q=repo%3Adotnet%2Fruntime%20CALLEE_DOES_NOT_RETURN&type=code) that makes codegen to think that the code after the method is unreachable and [replace it with int3](https://github.com/dotnet/runtime/blob/f5d53478febc328c42e4efcb43d0646d9be176cd/src/coreclr/jit/codegenlinear.cpp#L690-L693). RyuJIT should drop observations like this for methods outside the current version bubble.

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/95815

### CODE REVIEW

File: src/coreclr/vm/jitinterface.cpp [hunk 1]
```diff
@@ -6504,6 +6504,33 @@ bool CEEInfo::isIntrinsic(CORINFO_METHOD_HANDLE ftn)
...
+
+#ifdef PROFILING_SUPPORTED
+    if (CORProfilerEnableRejit())
+    {
+        // If ReJIT is around we'd better be careful with observations during inlining.
+        ret = false;
+    }
+#endif // PROFILING_SUPPORTED
+
+    EE_TO_JIT_TRANSITION_LEAF();
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:6529]:
```suggestion
    JIT_TO_EE_TRANSITION();

    _ASSERTE(ftn);

#ifdef FEATURE_REJIT
    MethodDesc *pCallee = GetMethod(ftn);
    MethodDesc *pCaller = m_pMethodBeingCompiled;
    pCallee->GetModule()->AddInlining(pCaller, pCallee);
#endif // PROFILING_SUPPORTED

    EE_TO_JIT_TRANSITION();
    
    return true;
```
Re-JIT can be enabled mid-flight. All code needs to be Re-JIT aware by default.

Comment (EgorBo) [reply src/coreclr/vm/jitinterface.cpp:6529]: Ah, yeah I had that in mind, but I thought that for the mid-flight case the only solution is to never do "no-return" logic in JIT at all.

File: src/coreclr/vm/jitinterface.cpp [hunk 2]
```diff
@@ -6504,6 +6504,29 @@ bool CEEInfo::isIntrinsic(CORINFO_METHOD_HANDLE ftn)
     return ret;
 }
 
+bool CEEInfo::notifyMethodInfoUsage(CORINFO_METHOD_HANDLE ftn)
+{
+    CONTRACTL {
+        NOTHROW;
+        GC_NOTRIGGER;
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:6511]:
```suggestion
        THROWS;
        GC_TRIGGERS;
```

**Comment by jkotas** [comment]
It would be nice to add a regression test for this. src\tests\readytorun\tests seems to be a good place - it is the only place that is testing opt-cross-module option.

**Comment by EgorBo** [comment]
> It would be nice to add a regression test for this. src\tests\readytorun\tests seems to be a good place - it is the only place that is testing opt-cross-module option.

Sure, will do. My understanding that the repro is:
1) Compile `ClassLibarary1.dll` with method `static void DoWork() => throw new Exception();`
2) Compile `ClassLibarary1.dll` with method `static void DoWork() => 42;`
3) Run crossgen for `ConsoleApp1` that invokes `DoWork()` with `--opt-cross-module="ClassLibrary1"`
4) Run resulting binary with `ClassLibarary1.dll` that returns 42.

File: src/coreclr/vm/jitinterface.cpp [hunk 3]
```diff
@@ -6504,6 +6504,29 @@ bool CEEInfo::isIntrinsic(CORINFO_METHOD_HANDLE ftn)
...
+
+    JIT_TO_EE_TRANSITION();
+
+    _ASSERTE(ftn);
+
+#ifdef FEATURE_REJIT
+    MethodDesc *pCallee = GetMethod(ftn);
+    MethodDesc *pCaller = m_pMethodBeingCompiled;
+    pCallee->GetModule()->AddInlining(pCaller, pCallee);
+#endif // PROFILING_SUPPORTED
```
Comment (AustinWise) [inline src/coreclr/vm/jitinterface.cpp:6523]: Nit: comment on `#endif` should match `#ifdef`:

```suggestion
#endif // FEATURE_REJIT
```

File: src/coreclr/inc/corinfo.h [hunk 4]
```diff
@@ -2050,6 +2050,8 @@ class ICorStaticInfo
     // Quick check whether the method is a jit intrinsic. Returns the same value as getMethodAttribs(ftn) & CORINFO_FLG_INTRINSIC, except faster.
     virtual bool isIntrinsic(CORINFO_METHOD_HANDLE ftn) = 0;
 
+    virtual bool notifyMethodInfoUsage(CORINFO_METHOD_HANDLE ftn) = 0;
```
Comment (jakobbotsch) [inline src/coreclr/inc/corinfo.h:2053]: It's not obvious to me what this returns -- add some documentation?

**Comment by EgorBo** [comment]
Added a test and verified that it fails on Main and is fixed in this PR

File: src/coreclr/inc/corinfo.h (hunk 4 above)
Comment (EgorBo) [reply src/coreclr/inc/corinfo.h:2053]: Sure, will add comments

**Comment by EgorBo** [comment]
@jkotas @jakobbotsch this is ready, added comments and a test (which fails on current Main)

Do we need to backport it to 8.0 since it was reported by a first party?

File: src/coreclr/inc/corinfo.h [hunk 5]
```diff
@@ -2050,6 +2050,19 @@ class ICorStaticInfo
...
+    // Notify EE about intent to rely on given MethodInfo in the current method
+    // EE returns false if we're not allowed to do so and the methodinfo may change.
+    // Example:
+    //  1) Crossgen (with --opt-cross-module=MyLib) attempts to inline a call from MyLib.dll into MyApp.dll
+    //     and realizes that the call is a throw helper.
+    //  2) JIT marks the call as no-return so it doesn't need an epilogue and is padded with ret3/brk
+    //     in the current method (MyApp.dll)
+    //  3) MyLib is updated to a new version so it's no longer within the same version bubble with MyApp.dll
+    //     and the new version of the call is no longer a throw helper and does some work.
+    //  4) ret3/brk is now reachable in the MyApp.dll.
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:2062]:
```suggestion
    // Example of a scenario addresses by notifyMethodInfoUsage:
    //  1) Crossgen (with --opt-cross-module=MyLib) attempts to inline a call from MyLib.dll into MyApp.dll
    //     and realizes that the call always throws.
    //  2) JIT aborts the inlining attempt and marks the call as no-return instead. The code that follows the call is 
    //     replaced with a breakpoint instruction that is expected to be unreachable.
    //  3) MyLib is updated to a new version so it's no longer within the same version bubble with MyApp.dll
    //     and the new version of the call no longer throws and does some work.
    //  4) The breakpoint instruction is now reachable in the MyApp.dll.
```

**Review by jkotas — APPROVED** [review]
Thank youy!

> Do we need to backport it to 8.0 since it was reported by a first party?

I do not think that this meets the bar for backport at this point. They have run into this bug because they use crossgen2 incorrectly. They need to fix how they use crossgen2.

File: src/coreclr/inc/corinfo.h [hunk 6]
```diff
@@ -2050,6 +2050,19 @@ class ICorStaticInfo
     // Quick check whether the method is a jit intrinsic. Returns the same value as getMethodAttribs(ftn) & CORINFO_FLG_INTRINSIC, except faster.
     virtual bool isIntrinsic(CORINFO_METHOD_HANDLE ftn) = 0;
 
+    // Notify EE about intent to rely on given MethodInfo in the current method
+    // EE returns false if we're not allowed to do so and the methodinfo may change.
+    // Example of a scenario addresses by notifyMethodInfoUsage:
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:2055]:
```suggestion
    // Example of a scenario addressed by notifyMethodInfoUsage:
```

**Pull request merged**